msg-socket/src/rep/driver.rs

```diff
@@ -291,62 +319,80 @@ impl<T: AsyncRead + AsyncWrite + Unpin, A: Address + Unpin> Stream for PeerState
             }
 
             // Then we check for completed requests, and push them onto the egress queue.
-            if let Poll::Ready(Some(Some((id, mut payload)))) =
+            if let Poll::Ready(Some(Spanned { inner: result, span })) =
                 this.pending_requests.poll_next_unpin(cx)
             {
-                let mut compression_type = 0;
-                let len_before = payload.len();
-                if let Some(ref compressor) = this.compressor {
-                    match compressor.compress(&payload) {
-                        Ok(compressed) => {
-                            payload = compressed;
-                            compression_type = compressor.compression_type() as u8;
-                        }
-                        Err(e) => {
-                            error!(err = ?e, "Failed to compress message");
-                            continue;
+                let span = span.entered();
+
+                match result {
+                    Err(_) => tracing::error!("response channel closed unexpectedly"),
+                    Ok(Response { msg_id, mut response }) => {
+                        let mut compression_type = 0;
+                        let len_before = response.len();
+                        if let Some(ref compressor) = this.compressor {
+                            match compressor.compress(&response) {
+                                Ok(compressed) => {
+                                    response = compressed;
+                                    compression_type = compressor.compression_type() as u8;
+                                }
+                                Err(e) => {
+                                    error!(?e, "failed to compress message");
+                                    continue;
+                                }
+                            }
+
+                            debug!(
+                                msg_id,
+                                len_before,
+                                len_after = response.len(),
+                                "compressed message"
+                            )
                         }
-                    }
 
-                    debug!(
-                        "Compressed message {} from {} to {} bytes",
-                        id,
-                        len_before,
-                        payload.len()
-                    )
-                }
+                        debug!(msg_id, "received response");
 
-                let msg = reqrep::Message::new(id, compression_type, payload);
-                this.egress_queue.push_back(msg);
+                        let msg = reqrep::Message::new(msg_id, compression_type, response);
+                        this.egress_queue.push_back(msg.with_span(span.clone()));
```

Comment:
also here, attaching an entered span. From what I understand the span should be entered/exiting when a certain piece of code is running, but shouldn't be passed around "entered". 
I mean, it would make any subsequent spans "child" of this one but this will also apply if the context is switched :S